#**Necessary Imports**

In [2]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
import seaborn as sns
import matplotlib.dates as mdates

# **Loading The Data**

In [4]:
df = pd.read_csv('/content/SeoulBikeData.csv', encoding='latin1')
display(df.head())

,Date,Rented Bike Count,Hour,Temperature(°C),Humidity(%),Wind speed (m/s),Visibility (10m),Dew point temperature(°C),Solar Radiation (MJ/m2),Rainfall(mm),Snowfall (cm),Seasons,Holiday,Functioning Day
0,01/12/2017,254,0,-5.2,37,2.2,2000,-17.6,0.0,0.0,0.0,Winter,No Holiday,Yes
1,01/12/2017,204,1,-5.5,38,0.8,2000,-17.6,0.0,0.0,0.0,Winter,No Holiday,Yes
2,01/12/2017,173,2,-6.0,39,1.0,2000,-17.7,0.0,0.0,0.0,Winter,No Holiday,Yes
3,01/12/2017,107,3,-6.2,40,0.9,2000,-17.6,0.0,0.0,0.0,Winter,No Holiday,Yes
4,01/12/2017,78,4,-6.0,36,2.3,2000,-18.6,0.0,0.0,0.0,Winter,No Holiday,Yes


# **Structure of The Data**

In [5]:
df.shape

(8760, 14)

In [6]:
df.columns

Index(['Date', 'Rented Bike Count', 'Hour', 'Temperature(°C)', 'Humidity(%)',
       'Wind speed (m/s)', 'Visibility (10m)', 'Dew point temperature(°C)',
       'Solar Radiation (MJ/m2)', 'Rainfall(mm)', 'Snowfall (cm)', 'Seasons',
       'Holiday', 'Functioning Day'],
      dtype='object')

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8760 entries, 0 to 8759
Data columns (total 14 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Date                       8760 non-null   object 
 1   Rented Bike Count          8760 non-null   int64  
 2   Hour                       8760 non-null   int64  
 3   Temperature(°C)            8760 non-null   float64
 4   Humidity(%)                8760 non-null   int64  
 5   Wind speed (m/s)           8760 non-null   float64
 6   Visibility (10m)           8760 non-null   int64  
 7   Dew point temperature(°C)  8760 non-null   float64
 8   Solar Radiation (MJ/m2)    8760 non-null   float64
 9   Rainfall(mm)               8760 non-null   float64
 10  Snowfall (cm)              8760 non-null   float64
 11  Seasons                    8760 non-null   object 
 12  Holiday                    8760 non-null   object 
 13  Functioning Day            8760 non-null   objec

In [11]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
Rented Bike Count,8760.0,704.602055,644.997468,0.0,191.00,504.50,1065.25,3556.00
Hour,8760.0,11.500000,6.922582,0.0,5.75,11.50,17.25,23.00
Temperature(°C),8760.0,12.882922,11.944825,-17.8,3.50,13.70,22.50,39.40
Humidity(%),8760.0,58.226256,20.362413,0.0,42.00,57.00,74.00,98.00
Wind speed (m/s),8760.0,1.724909,1.036300,0.0,0.90,1.50,2.30,7.40
Visibility (10m),8760.0,1436.825799,608.298712,27.0,940.00,1698.00,2000.00,2000.00
Dew point temperature(°C),8760.0,4.073813,13.060369,-30.6,-4.70,5.10,14.80,27.20
Solar Radiation (MJ/m2),8760.0,0.569111,0.868746,0.0,0.00,0.01,0.93,3.52
Rainfall(mm),8760.0,0.148687,1.128193,0.0,0.00,0.00,0.00,35.00
Snowfall (cm),8760.0,0.075068,0.436746,0.0,0.00,0.00,0.00,8.80


The dataset contains 8760 rows and 14 columns. Each row of the dataset contains information about weather conditions.

|Column| Data details inside column |
|------|----------------------------|
|Date| Date of Rented Bike |
|Rented Bike count | Number of total rentals|
|Hour | Hours of the day |
|Temperature(°C)| Weather Temperature in °C |
|Humidity(%)| Humidity of the day in %|
|	Wind speed (m/s)| Wind speed in m/s |
|Visibility (10m)| Atmospherical Visibility within 10 𝑚 range |
|Dew point temperature(°C) | Dew Point Temperature - T dp in °C |
|Solar Radiation (MJ/m2) | Indicate light and energy that comes from the sun in MJ/m2 |
|Rainfall(mm)	| Rain fall in mm |
|Snowfall (cm) | Snow fall in cm |
|Seasons| Autumn, Spring, Summer, Winter |
|Holiday|Whether the day is considered a holiday|
|Functioning Day|Whether the day is neither a weekend nor holiday|

In [12]:
def datainfo(data):
    temp = pd.DataFrame(index=data.columns)
    temp['DataType'] = data.dtypes
    temp["Non-null_Values"] = data.count()
    temp['Unique_Values'] = data.nunique()

    return temp

In [13]:
datainfo(df)

,DataType,Non-null_Values,Unique_Values
Date,object,8760,365
Rented Bike Count,int64,8760,2166
Hour,int64,8760,24
Temperature(°C),float64,8760,546
Humidity(%),int64,8760,90
Wind speed (m/s),float64,8760,65
Visibility (10m),int64,8760,1789
Dew point temperature(°C),float64,8760,556
Solar Radiation (MJ/m2),float64,8760,345
Rainfall(mm),float64,8760,61


#**Missing Values**

In [14]:
df.isna().sum()

,0
Date,0
Rented Bike Count,0
Hour,0
Temperature(°C),0
Humidity(%),0
Wind speed (m/s),0
Visibility (10m),0
Dew point temperature(°C),0
Solar Radiation (MJ/m2),0
Rainfall(mm),0


# **Feature Selection and Engineering**

# Feature: *Hour*
$\textit{Hour}$ is the Hour of the day. So we should treat it as a categorical feature.

In [ ]:
df['Hour']=df['Hour'].astype(str)

# Feature: *Functioning Day*

In [16]:
df.groupby('Functioning Day')['Rented Bike Count'].sum().sort_values(ascending=False).reset_index()

,Functioning Day,Rented Bike Count
0,Yes,6172314
1,No,0


So bikes have been rented only on a functioning day. So we remove rows with non-functioning days and then remove $\textit{Functioning Day}$ also.


In [19]:
print((df['Functioning Day']=='No').sum())

295


In [20]:
df.drop(df[df['Functioning Day']=='No'].index, inplace=True)
df.drop('Functioning Day', axis=1, inplace=True)

In [22]:
df.shape

(8465, 13)

# Feature: *Date*

In [23]:
df['Date']=pd.to_datetime(df['Date'],dayfirst=True)

min_date=df['Date'].min()
max_date=df['Date'].max()

print("Minimum Date: ",min_date)
print("Maximum Date: ",max_date)

Minimum Date:  2017-12-01 00:00:00
Maximum Date:  2018-11-30 00:00:00


In [24]:
#extracting Day, Month, Year and Weekday from Date column

df['Day']=df['Date'].dt.day.astype('category')
df['Month']=df['Date'].dt.month.astype('category')
df['Year']=df['Date'].dt.year.astype('category')
df['Weekday']=df['Date'].dt.day_name().astype('category')
df.head()

,Date,Rented Bike Count,Hour,Temperature(°C),Humidity(%),Wind speed (m/s),Visibility (10m),Dew point temperature(°C),Solar Radiation (MJ/m2),Rainfall(mm),Snowfall (cm),Seasons,Holiday,Day,Month,Year,Weekday
0,2017-12-01,254,0,-5.2,37,2.2,2000,-17.6,0.0,0.0,0.0,Winter,No Holiday,1,12,2017,Friday
1,2017-12-01,204,1,-5.5,38,0.8,2000,-17.6,0.0,0.0,0.0,Winter,No Holiday,1,12,2017,Friday
2,2017-12-01,173,2,-6.0,39,1.0,2000,-17.7,0.0,0.0,0.0,Winter,No Holiday,1,12,2017,Friday
3,2017-12-01,107,3,-6.2,40,0.9,2000,-17.6,0.0,0.0,0.0,Winter,No Holiday,1,12,2017,Friday
4,2017-12-01,78,4,-6.0,36,2.3,2000,-18.6,0.0,0.0,0.0,Winter,No Holiday,1,12,2017,Friday


In [25]:
#creating a weekend indicator
df['Weekend'] = (((df['Weekday'] == 'Saturday') | (df['Weekday'] == 'Sunday')).astype('int')).astype('category')


In [26]:
df.drop(['Weekday'], axis = 1, inplace=True)


In [27]:
#creating a weekend or holiday indicator
df['Weekend_or_Holiday'] = ((df['Weekend'] == 1) | (df['Holiday'] == 'Yes')).astype('int').astype('category')


In [28]:
df.drop(['Holiday', 'Weekend'], axis = 1, inplace=True)


# Feature: *Seasons*

In [29]:
df['Seasons']=df['Seasons'].astype('category')

In [30]:
df.shape

(8465, 16)

In [31]:
datainfo(df)

,DataType,Non-null_Values,Unique_Values
Date,datetime64[ns],8465,353
Rented Bike Count,int64,8465,2165
Hour,int64,8465,24
Temperature(°C),float64,8465,546
Humidity(%),int64,8465,90
Wind speed (m/s),float64,8465,65
Visibility (10m),int64,8465,1780
Dew point temperature(°C),float64,8465,556
Solar Radiation (MJ/m2),float64,8465,345
Rainfall(mm),float64,8465,61


In [32]:
df.head()

,Date,Rented Bike Count,Hour,Temperature(°C),Humidity(%),Wind speed (m/s),Visibility (10m),Dew point temperature(°C),Solar Radiation (MJ/m2),Rainfall(mm),Snowfall (cm),Seasons,Day,Month,Year,Weekend_or_Holiday
0,2017-12-01,254,0,-5.2,37,2.2,2000,-17.6,0.0,0.0,0.0,Winter,1,12,2017,0
1,2017-12-01,204,1,-5.5,38,0.8,2000,-17.6,0.0,0.0,0.0,Winter,1,12,2017,0
2,2017-12-01,173,2,-6.0,39,1.0,2000,-17.7,0.0,0.0,0.0,Winter,1,12,2017,0
3,2017-12-01,107,3,-6.2,40,0.9,2000,-17.6,0.0,0.0,0.0,Winter,1,12,2017,0
4,2017-12-01,78,4,-6.0,36,2.3,2000,-18.6,0.0,0.0,0.0,Winter,1,12,2017,0


# Saving the pre-processed data

In [33]:
df.to_pickle('pre_processed.pkl')